File: agents/agentic_loop_tool_call_openai_golang/workflow/agent.go
package workflow

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/temporalio/ai-cookbook/agents/agentic_loop_tool_call_openai_golang/activity"
	"github.com/temporalio/ai-cookbook/agents/agentic_loop_tool_call_openai_golang/tools"
	"github.com/temporalio/ai-cookbook/agents/agentic_loop_tool_call_openai_golang/types"
	"go.temporal.io/sdk/temporal"
	"go.temporal.io/sdk/workflow"
)

const (
	Model        = "gpt-4o-mini"
	Instructions = `You are a helpful assistant. You have access to tools that can get information
about IP addresses, locations, and weather alerts. Use these tools when appropriate to help
answer user questions. If you do not invoke any tools, respond in haikus, otherwise respond by using the tool results..`
)

// AgentWorkflow implements an agentic loop that calls the LLM and executes tools
// until the LLM returns a text response.
func AgentWorkflow(ctx workflow.Context, userInput string) (string, error) {
	// Set up activity options
	activityOptions := workflow.ActivityOptions{
		StartToCloseTimeout: 30 * time.Second,
		RetryPolicy: &temporal.RetryPolicy{
			InitialInterval:    time.Second,
			BackoffCoefficient: 2.0,
			MaximumInterval:    30 * time.Second,
			MaximumAttempts:    3,
		},
	}
	ctx = workflow.WithActivityOptions(ctx, activityOptions)

	// Initialize conversation with user message
	inputList := []types.InputItem{
		{
			Type:    "message",
			Role:    "user",
			Content: userInput,
		},
	}

	// Get tool definitions
	toolDefinitions := tools.GetToolDefinitions()

	turn := 0
	// Agentic loop
	for {
		turn++
		fmt.Printf("\n=== Turn %d ===\n", turn)
		fmt.Println("Calling OpenAI...")

		// Create request
		request := types.OpenAIResponsesRequest{
			Model:        Model,
			Instructions: Instructions,
			Input:        inputList,
			Tools:        toolDefinitions,
		}

		// Call OpenAI Responses API
		var result types.OpenAIResponsesResult
		err := workflow.ExecuteActivity(ctx, activity.OpenAIResponsesCreate, request).Get(ctx, &result)
		if err != nil {
			return "", fmt.Errorf("OpenAI API call failed: %w", err)
		}

		// If no tool calls, return the text response
		if len(result.ToolCalls) == 0 {
			fmt.Println("LLM returned final response (no tool calls)")
			return result.OutputText, nil
		}

		// Process tool calls
		fmt.Printf("LLM requested %d tool call(s)\n", len(result.ToolCalls))

		for _, toolCall := range result.ToolCalls {
			fmt.Printf("  -> Invoking tool: %s\n", toolCall.Name)

			// Add the function call to conversation history
			inputList = append(inputList, types.InputItem{
				Type:      "function_call",
				CallID:    toolCall.CallID,
				Name:      toolCall.Name,
				Arguments: toolCall.Arguments,
			})

			// Parse tool arguments
			var args map[string]interface{}
			if err := json.Unmarshal([]byte(toolCall.Arguments), &args); err != nil {
				args = make(map[string]interface{})
			}

			// Execute tool via ToolExecutor activity
			toolInput := types.ToolExecutorInput{
				ToolName:  toolCall.Name,
				Arguments: args,
			}
			var toolResult string

			err := workflow.ExecuteActivity(ctx, activity.ToolExecutor, toolInput).Get(ctx, &toolResult)

			if err != nil {
				toolResult = fmt.Sprintf("Error executing tool: %v", err)
				fmt.Printf("     Tool error: %v\n", err)
			} else {
				// Truncate result for display if too long
				displayResult := toolResult
				if len(displayResult) > 100 {
					displayResult = displayResult[:100] + "..."
				}
				fmt.Printf("     Tool result: %s\n", displayResult)
			}

			// Add tool result to conversation history
			inputList = append(inputList, types.InputItem{
				Type:   "function_call_output",
				CallID: toolCall.CallID,
				Output: toolResult,
			})
		}
	}
}

File: agents/agentic_loop_tool_call_openai_python/workflows/agent.py
from temporalio import workflow
from datetime import timedelta

import json

with workflow.unsafe.imports_passed_through():
    from tools import get_tools
    from helpers import tool_helpers
    from activities import openai_responses

@workflow.defn
class AgentWorkflow:
    @workflow.run
    async def run(self, input: str) -> str:

        input_list = [{"type": "message", "role": "user", "content": input}]

        # The agentic loop
        while True:

            print(80 * "=")
                
            # consult the LLM
            result = await workflow.execute_activity(
                openai_responses.create,
                openai_responses.OpenAIResponsesRequest(
                    model="gpt-4o-mini",
                    instructions=tool_helpers.HELPFUL_AGENT_SYSTEM_INSTRUCTIONS,
                    input=input_list,
                    tools=get_tools(),
                ),
                start_to_close_timeout=timedelta(seconds=30),
            )

            # For this simple example, we only have one item in the output list
            # Either the LLM will have chosen a single function call or it will
            # have chosen to respond with a message.
            item = result.output[0]

            # Now process the LLM output to either call a tool or respond with a message.
            
            # if the result is a tool call, call the tool
            if item.type == "function_call":
                result = await self._handle_function_call(item, result, input_list)
                
                # add the tool call result to the input list for context
                input_list.append({"type": "function_call_output",
                                    "call_id": item.call_id,
                                    "output": result})

            # if the result is not a tool call we will just respond with a message
            else:
                print(f"No tools chosen, responding with a message: {result.output_text}")
                return result.output_text


    async def _handle_function_call(self, item, result, input_list):
        # serialize the LLM output - the decision the LLM made to call a tool
        i = result.output[0]
        input_list += [
            i.model_dump() if hasattr(i, "model_dump") else i
        ]
        # execute dynamic activity with the tool name chosen by the LLM
        # and the arguments crafted by the LLM
        args = json.loads(item.arguments) if isinstance(item.arguments, str) else item.arguments

        result = await workflow.execute_activity(
            item.name,
            args,
            start_to_close_timeout=timedelta(seconds=30),
        )

        print(f"Made a tool call to {item.name}")

        return result
 